File: pkg/services/poster_service.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"cloud.google.com/go/storage"
)

const (
	tmdbSearchURL = "https://api.themoviedb.org/3/search/movie"
	tmdbImageBase = "https://image.tmdb.org/t/p/w500"
)

// TMDbSearchResult represents a movie search result from TMDb
type TMDbSearchResult struct {
	Results []struct {
		ID          int     `json:"id"`
		Title       string  `json:"title"`
		PosterPath  *string `json:"poster_path"`
		ReleaseDate string  `json:"release_date"`
	} `json:"results"`
}

// FetchMoviePoster searches for a movie poster and uploads it to storage
func (s *Service) FetchMoviePoster(videoPath string, movieTitle string, progressCb ProgressCallback) error {
	sendProgress := func(step string, progress int) {
		if progressCb != nil {
			progressCb(step, progress)
		}
	}

	sendProgress("Getting API key", 5)
	apiKey := os.Getenv("TMDB_API_KEY")
	if apiKey == "" {
		return fmt.Errorf("TMDB_API_KEY environment variable not set")
	}

	sendProgress("Searching for movie", 15)

	// Clean the movie title for better search results
	cleanTitle := cleanMovieTitle(movieTitle)

	// Search for movie
	searchURL := fmt.Sprintf("%s?api_key=%s&query=%s", tmdbSearchURL, apiKey, url.QueryEscape(cleanTitle))
	resp, err := http.Get(searchURL)
	if err != nil {
		return fmt.Errorf("failed to search movie: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("TMDb API error (status %d): %s", resp.StatusCode, string(body))
	}

	var result TMDbSearchResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return fmt.Errorf("failed to decode search result: %v", err)
	}

	if len(result.Results) == 0 {
		return fmt.Errorf("no movie found for title: %s", movieTitle)
	}

	// Try to find exact match first, then fall back to partial match
	movie := findBestMatch(result.Results, cleanTitle)
	if movie.PosterPath == nil || *movie.PosterPath == "" {
		return fmt.Errorf("no poster available for: %s", movieTitle)
	}

	sendProgress("Downloading poster", 40)

	// Download poster image
	posterURL := tmdbImageBase + *movie.PosterPath
	posterResp, err := http.Get(posterURL)
	if err != nil {
		return fmt.Errorf("failed to download poster: %v", err)
	}
	defer posterResp.Body.Close()

	if posterResp.StatusCode != http.StatusOK {
		return fmt.Errorf("failed to download poster (status %d)", posterResp.StatusCode)
	}

	sendProgress("Saving poster", 70)

	// Save to temp file
	outputDir := filepath.Join(os.TempDir(), "video-gallery-thumbnails")
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return fmt.Errorf("failed to create output directory: %v", err)
	}

	ext := filepath.Ext(videoPath)
	basePath := videoPath[:len(videoPath)-len(ext)]
	thumbnailPath := basePath + ".jpg"

	thumbnailBaseName := getSafeFilename(thumbnailPath)
	tmpThumbnailPath := filepath.Join(outputDir, thumbnailBaseName)
	cleanTmpPath := filepath.Clean(tmpThumbnailPath)
	if !strings.HasPrefix(cleanTmpPath, filepath.Clean(outputDir)) {
		return fmt.Errorf("invalid path: path traversal detected")
	}

	tmpFile, err := os.Create(cleanTmpPath)
	if err != nil {
		return fmt.Errorf("failed to create temp file: %v", err)
	}
	defer tmpFile.Close()
	defer os.Remove(cleanTmpPath)

	if _, err := io.Copy(tmpFile, posterResp.Body); err != nil {
		return fmt.Errorf("failed to save poster: %v", err)
	}
	tmpFile.Close()

	sendProgress("Uploading to storage", 85)

	// Upload to cloud storage
	ctx := context.Background()
	client, err := storage.NewClient(ctx)
	if err != nil {
		return fmt.Errorf("failed to create storage client: %v", err)
	}
	defer client.Close()

	bucket := client.Bucket(s.config.BucketName)

	// Clear old thumbnail
	bucket.Object(thumbnailPath).Delete(ctx)

	if err := uploadFile(ctx, bucket, cleanTmpPath, thumbnailPath); err != nil {
		return fmt.Errorf("error uploading poster: %v", err)
	}

	sendProgress("Clearing cache", 95)
	s.videoCache.Flush()

	sendProgress("Complete", 100)
	log.Printf("Successfully fetched poster for: %s", movieTitle)
	return nil
}

// SearchMoviePoster searches for a movie and returns available posters
func (s *Service) SearchMoviePoster(movieTitle string) ([]MoviePosterResult, error) {
	apiKey := os.Getenv("TMDB_API_KEY")
	if apiKey == "" {
		return nil, fmt.Errorf("TMDB_API_KEY environment variable not set")
	}

	// Clean the movie title for better search results
	cleanTitle := cleanMovieTitle(movieTitle)

	searchURL := fmt.Sprintf("%s?api_key=%s&query=%s", tmdbSearchURL, apiKey, url.QueryEscape(cleanTitle))
	resp, err := http.Get(searchURL)
	if err != nil {
		return nil, fmt.Errorf("failed to search movie: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("TMDb API error (status %d): %s", resp.StatusCode, string(body))
	}

	var result TMDbSearchResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode search result: %v", err)
	}

	var posters []MoviePosterResult
	for _, movie := range result.Results {
		if movie.PosterPath != nil && *movie.PosterPath != "" {
			posters = append(posters, MoviePosterResult{
				Title:        movie.Title,
				Year:         extractYear(movie.ReleaseDate),
				PosterURL:    tmdbImageBase + *movie.PosterPath,
				ThumbnailURL: strings.Replace(tmdbImageBase, "w500", "w185", 1) + *movie.PosterPath,
			})
		}
	}

	return posters, nil
}

// MoviePosterResult represents a movie poster search result
type MoviePosterResult struct {
	Title        string `json:"title"`
	Year         string `json:"year"`
	PosterURL    string `json:"posterUrl"`
	ThumbnailURL string `json:"thumbnailUrl"`
}

func extractYear(releaseDate string) string {
	if len(releaseDate) >= 4 {
		return releaseDate[:4]
	}
	return ""
}

// findBestMatch finds the best matching movie from results
// First tries exact match (case-insensitive), then falls back to partial match
func findBestMatch(results []struct {
	ID          int     `json:"id"`
	Title       string  `json:"title"`
	PosterPath  *string `json:"poster_path"`
	ReleaseDate string  `json:"release_date"`
}, searchTitle string) struct {
	ID          int     `json:"id"`
	Title       string  `json:"title"`
	PosterPath  *string `json:"poster_path"`
	ReleaseDate string  `json:"release_date"`
} {
	searchLower := strings.ToLower(searchTitle)

	// First pass: look for exact match
	for _, movie := range results {
		if strings.ToLower(movie.Title) == searchLower {
			return movie
		}
	}

	// Second pass: look for partial match (contains)
	for _, movie := range results {
		if strings.Contains(strings.ToLower(movie.Title), searchLower) {
			return movie
		}
	}

	// No match found, return first result
	return results[0]
}

// cleanMovieTitle removes common metadata from movie titles for better search results
// Examples: "Empire Strikes Back (Despecialized v2 0)" -> "Empire Strikes Back"
func cleanMovieTitle(title string) string {
	// Remove content in parentheses (e.g., version info, year, quality)
	if idx := strings.Index(title, "("); idx != -1 {
		title = title[:idx]
	}

	// Remove content in brackets
	if idx := strings.Index(title, "["); idx != -1 {
		title = title[:idx]
	}

	// Trim whitespace
	title = strings.TrimSpace(title)

	return title
}

File: src/services/poster_service.rs
use crate::config::Config;
use crate::models::MoviePosterResult;
use crate::services::thumbnail_service::ProgressCallback;
use anyhow::{anyhow, Result};
use cloud_storage::Client;
use regex::Regex;
use serde::Deserialize;
use std::path::Path;
use std::sync::Arc;
use tokio::fs;
use tokio::io::AsyncWriteExt;
use tracing::{debug, error, info};

#[derive(Debug, Deserialize)]
struct TMDbSearchResponse {
    results: Vec<TMDbMovie>,
}

#[derive(Debug, Deserialize)]
struct TMDbMovie {
    title: String,
    release_date: Option<String>,
    poster_path: Option<String>,
}

#[derive(Clone)]
pub struct PosterService {
    config: Arc<Config>,
    client: reqwest::Client,
}

impl PosterService {
    pub fn new(config: Arc<Config>) -> Self {
        Self {
            config,
            client: reqwest::Client::new(),
        }
    }

    pub async fn search_movie_poster(&self, movie_title: &str) -> Result<Vec<MoviePosterResult>> {
        let api_key = self
            .config
            .tmdb_api_key
            .as_ref()
            .ok_or_else(|| anyhow!("TMDB_API_KEY not configured"))?;

        let url = format!(
            "https://api.themoviedb.org/3/search/movie?api_key={}&query={}",
            api_key,
            urlencoding::encode(movie_title)
        );

        let response = self
            .client
            .get(&url)
            .send()
            .await
            .map_err(|e| anyhow!("Failed to search TMDb: {}", e))?;

        let search_result: TMDbSearchResponse = response
            .json()
            .await
            .map_err(|e| anyhow!("Failed to parse TMDb response: {}", e))?;

        let results = search_result
            .results
            .into_iter()
            .map(|movie| {
                let year = movie
                    .release_date
                    .as_ref()
                    .and_then(|date| date.split('-').next())
                    .and_then(|year_str| year_str.parse::<i32>().ok());

                MoviePosterResult {
                    title: movie.title,
                    year,
                    poster_path: movie.poster_path,
                }
            })
            .collect();

        Ok(results)
    }

    pub async fn fetch_movie_poster(
        &self,
        video_path: &str,
        movie_title: &str,
        progress_cb: Option<ProgressCallback>,
    ) -> Result<()> {
        let send_progress = |step: String, progress: i32| {
            if let Some(cb) = &progress_cb {
                cb(step, progress);
            }
        };

        send_progress("Searching for movie".to_string(), 10);

        // Extract movie title from filename if empty
        let title = if movie_title.is_empty() {
            extract_movie_title(video_path)
        } else {
            movie_title.to_string()
        };

        info!("Searching for movie: {}", title);

        let results = self.search_movie_poster(&title).await?;
        
        if results.is_empty() {
            return Err(anyhow!("No movies found for: {}", title));
        }

        let movie = &results[0];
        let poster_path = movie
            .poster_path
            .as_ref()
            .ok_or_else(|| anyhow!("No poster available for: {}", title))?;

        send_progress("Downloading poster".to_string(), 40);

        let poster_url = format!("https://image.tmdb.org/t/p/w500{}", poster_path);
        let poster_data = self
            .client
            .get(&poster_url)
            .send()
            .await
            .map_err(|e| anyhow!("Failed to download poster: {}", e))?
            .bytes()
            .await
            .map_err(|e| anyhow!("Failed to read poster data: {}", e))?;

        send_progress("Uploading poster to storage".to_string(), 70);

        // Generate thumbnail path
        let thumbnail_path = Self::get_thumbnail_path(video_path);

        let storage_client = Client::default();
        
        // Delete old thumbnail if it exists
        let _ = storage_client
            .object()
            .delete(&self.config.bucket_name, &thumbnail_path)
            .await;

        // Upload poster as thumbnail
        storage_client
            .object()
            .create(
                &self.config.bucket_name,
                poster_data.to_vec(),
                &thumbnail_path,
                "image/jpeg",
            )
            .await
            .map_err(|e| anyhow!("Failed to upload poster: {}", e))?;

        send_progress("Complete".to_string(), 100);
        info!("Successfully fetched poster for {}", title);

        Ok(())
    }

    fn get_thumbnail_path(video_path: &str) -> String {
        let path = Path::new(video_path);
        let stem = path.file_stem().unwrap_or_default();
        let parent = path.parent().unwrap_or_else(|| Path::new(""));
        parent.join(stem).with_extension("jpg").to_string_lossy().to_string()
    }
}

fn extract_movie_title(filename: &str) -> String {
    // Remove path and extension
    let path = Path::new(filename);
    let name = path
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or(filename);

    // Remove year in brackets or parentheses
    let year_regex = Regex::new(r"[\[\(]\d{4}[\]\)]").unwrap();
    let name = year_regex.replace_all(name, "");

    // Replace common separators with spaces
    let name = name.replace(['.', '_', '-'], " ");

    // Remove extra whitespace
    let name = name.split_whitespace().collect::<Vec<_>>().join(" ");

    name.trim().to_string()
}
